### PR TITLE
Enhancement/59 rework column metadata

### PR DESCRIPTION
**Pull request by redcatbear**

### CODE REVIEW

File: src/main/java/com/exasol/adapter/metadata/DataType.java [hunk 1]
```diff
@@ -54,6 +43,16 @@ public static DataType createVarChar(final int size, final ExaCharset charset) {
         return type;
     }
 
+    /**
+     * Create a <code>VARCHAR</code> data type with the maximum size supported by Exasol
+     *
+     * @param charset character set to be used when creating the data type
+     * @return data type
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/metadata/DataType.java]: return `varchar` data type.

same for others below.

File: src/main/java/com/exasol/adapter/metadata/DataType.java [hunk 2]
```diff
@@ -54,6 +43,16 @@ public static DataType createVarChar(final int size, final ExaCharset charset) {
...
+     */
+    public static DataType createMaximumSizeVarChar(final ExaCharset charset) {
+        return createVarChar(MAX_EXASOL_VARCHAR_SIZE, charset);
+    }
+
     public static DataType createChar(final int size, final ExaCharset charset) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/metadata/DataType.java]: Add Javadoc to public method or make it private method.

File: src/main/java/com/exasol/adapter/metadata/DataType.java [hunk 3]
```diff
@@ -62,6 +61,16 @@ public static DataType createChar(final int size, final ExaCharset charset) {
...
+     * Create a <code>CHAR</code> data type with the maximum size supported by Exasol
+     *
+     * @param charset character set to be used when creating the data type
+     * @return data type
+     */
+    public static DataType createMaximumSizeChar(final ExaCharset charset) {
+        return createChar(MAX_EXASOL_CHAR_SIZE, charset);
+    }
+
     public static DataType createDecimal(final int precision, final int scale) {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/metadata/DataType.java]: Add Javadoc or private.

File: src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java [hunk 4]
```diff
@@ -1,21 +1,25 @@
...
-import static com.exasol.adapter.request.parser.RequestParserConstants.*;
-
+import com.exasol.adapter.metadata.ColumnMetadata;
+import com.exasol.adapter.metadata.DataType;
+import com.exasol.adapter.metadata.TableMetadata;
+
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.json.JsonString;
+import javax.json.JsonValue;
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java]: More than three use star import.

File: src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java [hunk 5]
```diff
@@ -77,35 +82,35 @@ private DataType getDataType(final JsonObject dataType) {
...
-            break;
-        case "INTERVAL":
-            type = getIntervalDataType(dataType);
-            break;
-        case "GEOMETRY":
-            type = getGeometryDataType(dataType);
-            break;
-        default:
-            throw new RequestParserException("Unsupported data type encountered: " + typeName);
+            case "DECIMAL":
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java]: This is not consistent formatting with above another class with `switch`.

Please create a ticket for having unified formatter.

File: src/test/java/com/exasol/adapter/AdapterPropertiesTest.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    void testGetFilteredTables() {
+        this.rawProperties.put(AdapterProperties.TABLE_FILTER_PROPERTY, "Table a,Table B, TABLE  C    ,  table d  ");
+        final AdapterProperties properties = createProperties();
+        assertThat(properties.getFilteredTables(), containsInAnyOrder("Table a", "Table B", "TABLE  C", "table d"));
+    }
+
+    @ValueSource(strings = { CONNECTION_STRING_PROPERTY, CONNECTION_NAME_PROPERTY, USERNAME_PROPERTY, PASSWORD_PROPERTY,
+            SCHEMA_NAME_PROPERTY, CATALOG_NAME_PROPERTY, TABLE_FILTER_PROPERTY })
+    @ParameterizedTest
+    void testtIsRefreshingVirtualSchemaRequiredTrue(final String propertyName) {
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/AdapterPropertiesTest.java]: Typo, `testt` to `test`

File: src/test/java/com/exasol/adapter/AdapterPropertiesTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        this.rawProperties.put(propertyName, "");
+        assertThat(AdapterProperties.isRefreshingVirtualSchemaRequired(this.rawProperties), equalTo(true));
+    }
+
+    @Test
+    void testtIsRefreshingVirtualSchemaRequiredFalse() {
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/AdapterPropertiesTest.java]: Typo, `testt` to `test`

File: src/test/java/com/exasol/adapter/json/SchemaMetadataSerializerTest.java [hunk 8]
```diff
@@ -1,21 +1,22 @@
...
-
-import java.util.ArrayList;
-import java.util.List;
-
-import javax.json.JsonObjectBuilder;
-
+import com.exasol.adapter.metadata.ColumnMetadata;
+import com.exasol.adapter.metadata.DataType;
+import com.exasol.adapter.metadata.SchemaMetadata;
+import com.exasol.adapter.metadata.TableMetadata;
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/json/SchemaMetadataSerializerTest.java]: Use star import, more than four.

File: src/test/java/com/exasol/adapter/request/parser/PushDownSqlParserTest.java [hunk 9]
```diff
@@ -90,7 +92,7 @@ void testParseSelectWithLimit() {
         }
         final SqlStatementSelect sqlStatementSelect =
               (SqlStatementSelect) this.pushdownSqlParser.parseExpression(this.jsonObject);
-        final SqlLimit sqlLimit = (SqlLimit) sqlStatementSelect.getLimit();
+        final SqlLimit sqlLimit = sqlStatementSelect.getLimit();
```
Comment (morazow) [inline src/test/java/com/exasol/adapter/request/parser/PushDownSqlParserTest.java]: :+1:

File: src/main/java/com/exasol/adapter/AbstractAdapterProperties.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+    public boolean isEnabled(final String key) {
+        return "true".equalsIgnoreCase(this.properties.get(key));
+    }
+
+    /**
+     * Get value for key
+     *
+     * @param key property to search for
+     * @return corresponding value or null it the key-value pair does not exist
+     */
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/AbstractAdapterProperties.java]: Fixed.

File: src/main/java/com/exasol/adapter/AbstractAdapterProperties.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    public boolean containsKey(final String key) {
+        return this.properties.containsKey(key);
+    }
+
+    /**
+     * Check if the switch with the given key is enabled
+     *
+     * @param key switch name
+     * @return <code>true</code> if the switch property exists and is enable
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/AbstractAdapterProperties.java]: Fixed.

File: src/main/java/com/exasol/adapter/AbstractAdapterProperties.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+     */
+    public boolean isEnabled(final String key) {
+        return "true".equalsIgnoreCase(this.properties.get(key));
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/AbstractAdapterProperties.java]: `get()` returns `null` if the key does not exist. This results in the method returning `false`.

File: src/main/java/com/exasol/adapter/AdapterProperties.java [hunk 13]
```diff
@@ -0,0 +1,193 @@
+package com.exasol.adapter;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class AdapterProperties extends AbstractAdapterProperties {
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/AdapterProperties.java]: Added. Please review.

File: src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java (hunk 5 above)
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/request/parser/TablesMetadataParser.java]: Delegated to @AnastasiiaSergienko.

File: src/test/java/com/exasol/adapter/AdapterPropertiesTest.java (hunk 6 above)
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/AdapterPropertiesTest.java]: Fixed.

File: src/test/java/com/exasol/adapter/AdapterPropertiesTest.java (hunk 7 above)
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/AdapterPropertiesTest.java]: Fixed.

File: src/test/java/com/exasol/adapter/json/SchemaMetadataSerializerTest.java (hunk 8 above)
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/json/SchemaMetadataSerializerTest.java]: For members we do this at three or more, but I would not do that for classes.

File: src/test/java/com/exasol/adapter/request/parser/PushDownSqlParserTest.java (hunk 9 above)
Comment (redcatbear) [reply src/test/java/com/exasol/adapter/request/parser/PushDownSqlParserTest.java]: :)

File: src/main/java/com/exasol/adapter/metadata/ColumnMetadata.java [hunk 14]
```diff
@@ -5,87 +5,304 @@
...
     private final String adapterNotes;
     private final DataType type;
-    private final boolean isNullable;
-    private final boolean isIdentity;
+    private final boolean nullable;
+    private final boolean identity;
     private final String defaultValue;
+    private final boolean defaultValueExplicitlySet;
     private final String comment;
+    private final String orginialTypeName;
```
[the remark this answers is not in the document]
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/metadata/ColumnMetadata.java]: Fixed.

File: src/main/java/com/exasol/adapter/metadata/DataType.java (hunk 1 above)
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/metadata/DataType.java]: Done.

File: src/main/java/com/exasol/adapter/metadata/DataType.java (hunk 3 above)
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/metadata/DataType.java]: Done.

File: src/main/java/com/exasol/adapter/metadata/DataType.java (hunk 2 above)
Comment (redcatbear) [reply src/main/java/com/exasol/adapter/metadata/DataType.java]: Done.

**Comment by morazow** [comment]
LGTM :+1:

**Pull request merged**